packages/ui/src/containers/Passwordless/EmailPasswordless.tsx

```diff
@@ -0,0 +1,157 @@
+/**
+ * TODO:
+ * 1. API redesign handle api error and loading status globally in PageContext
+ * 2. Input field validation, should move the validation rule to the input field scope
+ * 4. Read terms of use settings from SignInExperience Settings
+ */
+import { LogtoErrorI18nKey } from '@logto/phrases';
+import classNames from 'classnames';
+import React, { useState, useCallback, useMemo, useEffect, useContext } from 'react';
+import { useTranslation } from 'react-i18next';
+
+import { sendEmailPasscode as sendRegisterEmailPasscode } from '@/apis/register';
+import { sendEmailPasscode as sendSignInEmailPasscode } from '@/apis/sign-in';
+import Button from '@/components/Button';
+import { ErrorType } from '@/components/ErrorMessage';
+import Input from '@/components/Input';
+import TermsOfUse from '@/components/TermsOfUse';
+import PageContext from '@/hooks/page-context';
+import useApi from '@/hooks/use-api';
+
+import * as styles from './index.module.scss';
+
+type Props = {
+  type: 'sign-in' | 'register';
+};
+
+type FieldState = {
+  email: string;
+  termsAgreement: boolean;
+};
+
+type ErrorState = {
+  [key in keyof FieldState]?: ErrorType;
+};
+
+type FieldValidations = {
+  [key in keyof FieldState]: (state: FieldState) => ErrorType | undefined;
+};
+
+const defaultState: FieldState = { email: '', termsAgreement: false };
+
+const emailRegEx = /^\S+@\S+\.\S+$/;
+
+const EmailPasswordless = ({ type }: Props) => {
+  const { t, i18n } = useTranslation();
+  const [fieldState, setFieldState] = useState<FieldState>(defaultState);
+  const [fieldErrors, setFieldErrors] = useState<ErrorState>({});
+  const { setToast } = useContext(PageContext);
+
+  const sendPasscode = type === 'sign-in' ? sendSignInEmailPasscode : sendRegisterEmailPasscode;
+
+  const { loading, error, result, run: asyncSendPasscode } = useApi(sendPasscode);
+
+  const validations = useMemo<FieldValidations>(
```

Comment:
nit: I am more familiar with 2 useCallbacks.

Reply:
Truly not much difference in this case. If you check the next PR #457 which contains up to 4 fields that need to be validated. I do it in this way so I can loop through each field validation method easily.  like this:


<img width="573" alt="image" src="https://user-images.githubusercontent.com/36393111/160809688-4ff366c6-5792-41cf-9387-bcfd74a4fc45.png">

Reply:
prevent bunch of if else

Reply:
@wangsijie let me know if you have other preferences thx